src/Components/Web.JS/src/Rendering/Events/EventTypes.ts

```diff
@@ -0,0 +1,377 @@
+interface EventTypeOptions {
+  browserEventName?: string;
+  createEventArgs?: (event: Event) => any;
+}
+
+const eventTypeRegistry: Map<string, EventTypeOptions> = new Map();
+const browserEventNamesToAliases: Map<string, string[]> = new Map();
+const createBlankEventArgsOptions: EventTypeOptions = { createEventArgs: () => ({}) };
+
+export const eventNameAliasRegisteredCallbacks: ((aliasEventName: string, browserEventName) => void)[] = [];
+
+export function registerCustomEventType(eventName: string, options: EventTypeOptions): void {
+  if (!options) {
+    throw new Error('The options parameter is required.');
+  }
+
+  // There can't be more than one registration for the same event name because then we wouldn't
+  // know which eventargs data to supply.
+  if (eventTypeRegistry.has(eventName)) {
+    throw new Error(`The event '${eventName}' is already registered.`);
+  }
+
+  // If applicable, register this as an alias of the given browserEventName
+  if (options.browserEventName) {
+    const aliasGroup = browserEventNamesToAliases.get(options.browserEventName);
+    if (aliasGroup) {
+      aliasGroup.push(eventName);
+    } else {
+      browserEventNamesToAliases.set(options.browserEventName, [eventName]);
+    }
+
+    // For developer convenience, it's allowed to register the custom event type *after*
+    // some listeners for it are already present. Once the event name alias gets registered,
+    // we have to notify any existing event delegators so they can update their delegated
+    // events list.
+    eventNameAliasRegisteredCallbacks.forEach(callback => callback(eventName, options.browserEventName));
```

Comment:
Can you give an example on when you think this is useful?

Reply:
Sure. If we didn't do this, we'd be relying on the event handlers being registered on the JS side before the corresponding elements get added to the DOM. This would be really inconvenient because you'd now be forced to use the `autostart=false` thing on `<script src="blazor.*.js">` so you could be sure to register the event handler before the first render. It would be worse still for component library authors, as they'd have to do some kind of deferred async rendering flow to guarantee their event handler registrations go through before their component(s) first appear in the UI.

I judged that these usability issues were unacceptable and needed a solution. So with the mechanism you see here, the event handler can be registered after the elements are rendered in the DOM. The only requirement is that the event handler is registered before the event actually occurs (i.e., in response to a user action), of course.